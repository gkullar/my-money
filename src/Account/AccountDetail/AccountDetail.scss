.account-detail {
  display: flex;
  justify-content: space-between;
  align-items: center;

  .account-detail-title {
    font-weight: bold;
    font-size: 14px;
    margin-right: 5px;
  }

  .account-detail-value {
    font-size: 22px;
  }
}
