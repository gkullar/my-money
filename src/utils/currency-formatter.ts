export const toGBP = (value: number): string => 'Â£' + (value / 100).toFixed(2);
