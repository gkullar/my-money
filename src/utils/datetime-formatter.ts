export const toDateTime = (value: string): string =>
  new Date(value).toLocaleString();
