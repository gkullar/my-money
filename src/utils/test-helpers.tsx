import { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import 'jest-styled-components';
import AppProviders from '../app-providers';

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AppProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
