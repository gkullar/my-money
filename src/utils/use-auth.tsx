import React, {
  createContext,
  useContext,
  useState,
  FunctionComponent
} from 'react';
import uuid from './uuid';

const accessTokenKey = `${process.env.REACT_APP_API_ACCESS_TOKEN_KEY}`;
const stateTokenKey = `${process.env.REACT_APP_API_STATE_TOKEN_KEY}`;
const clientId = `${process.env.REACT_APP_API_CLIENT_ID}`;
const clientSecret = `${process.env.REACT_APP_API_CLIENT_SECRET}`;
const redirectUri = 'http://localhost:3000'; // app not for production!

if (!localStorage.getItem(stateTokenKey))
  localStorage.setItem(stateTokenKey, uuid());

const stateToken = localStorage.getItem(stateTokenKey);

interface State {
  isAuthenticated: boolean;
  login: () => void;
  logout: () => void;
}

const authContext = createContext<State>({} as any);

const useAuth = () => useContext(authContext);

const AuthProvider: FunctionComponent<{}> = ({ children }) => {
  const auth = useAuthProvider();
  return <authContext.Provider value={auth}>{children}</authContext.Provider>;
};

function useAuthProvider(): State {
  const [isAuthenticated, setIsAuthenticated] = useState<boolean>(
    localStorage.getItem(accessTokenKey) ? true : false
  );
  const params = new URLSearchParams(window.location.search);

  if (!isAuthenticated && params.has('code') && params.has('state')) {
    if (params.get('state') !== stateToken)
      throw new Error('Error: state token not matched'); // @todo handle error in UI

    authenticate(params.get('code') as string);
  }

  async function authenticate(code: string) {
    const data = new FormData();
    data.append('grant_type', 'authorization_code');
    data.append('client_id', clientId);
    data.append('client_secret', clientSecret);
    data.append('redirect_uri', redirectUri);
    data.append('code', code);

    const response = await fetch(
      `${process.env.REACT_APP_API_URL}/oauth2/token`,
      {
        method: 'post',
        body: data
      }
    );

    const result = await response.json();

    if (response.status !== 200 || result.error)
      throw new Error('Error: Failed Authentication'); // @todo handle error in UI

    localStorage.setItem(accessTokenKey, result.access_token);
    setIsAuthenticated(true);
    window.history.replaceState(null, document.title, window.location.pathname); // remove query params
  }

  function login() {
    window.location.href = `https://auth.monzo.com/?client_id=${clientId}&redirect_uri=${redirectUri}&response_type=code&state=${stateToken}`;
  }

  async function logout() {
    const response = await fetch(
      `${process.env.REACT_APP_API_URL}/oauth2/logout`,
      {
        method: 'post',
        headers: new Headers({
          Authorization: `Bearer ${localStorage.getItem(accessTokenKey)}`
        })
      }
    );

    const result = await response.json();

    if (!result.ok) throw new Error('Error: Failed to logout'); // @todo handle error in UI

    localStorage.removeItem(accessTokenKey);
    setIsAuthenticated(false);
  }

  return { isAuthenticated, login, logout };
}

export { useAuth, AuthProvider };
