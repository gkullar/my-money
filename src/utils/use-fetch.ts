import { useState, useEffect } from 'react';

const accessTokenKey = `${process.env.REACT_APP_API_ACCESS_TOKEN_KEY}`;

interface State<T> {
  data: T;
  loading: boolean;
  hasErrors: boolean;
  hasPermissions: boolean;
}

const useFetch = <T>(url: string): State<T> => {
  const [data, setData] = useState<T>({} as T);
  const [loading, setLoading] = useState<boolean>(true);
  const [hasErrors, setHasErrors] = useState<boolean>(false);
  const [hasPermissions, setHasPermissions] = useState<boolean>(true);

  async function fetchUrl() {
    const response = await fetch(url, {
      method: 'get',
      headers: new Headers({
        Authorization: `Bearer ${localStorage.getItem(accessTokenKey)}`
      })
    });

    const result = await response.json();

    if (response.status === 401) {
      localStorage.removeItem(accessTokenKey);
      window.location.href = '/';
    } else if (response.status === 403) {
      setHasPermissions(false);
    } else if (response.status !== 200) {
      throw new Error('Error: Failed Data Fetch'); // @todo handle error in UI
    } else {
      setData(result);
      setLoading(false);
      setHasErrors(false);
      setHasPermissions(true);
    }
  }

  useEffect(() => {
    fetchUrl();
  }, []);

  return { data, loading, hasErrors, hasPermissions };
};

export default useFetch;
