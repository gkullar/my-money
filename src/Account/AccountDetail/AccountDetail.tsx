import React from 'react';
import './AccountDetail.scss';

interface Props {
  title: string;
  value: number;
}

export default function AccountDetail(props: Props) {
  return (
    <div className="account-detail">
      <span className="title">{props.title}</span>
      <span className="value">
        <small>Â£</small>
        {(props.value / 100).toFixed(2)}
      </span>
    </div>
  );
}
