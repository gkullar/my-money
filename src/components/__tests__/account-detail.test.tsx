import React from 'react';
import AccountDetail from '../account-detail';
import { render } from '../../utils/test-helpers';

describe('AccountDetail', () => {
  test('renders correctly', () => {
    const { container, getByText } = render(
      <AccountDetail title="TestTitle" value="TestValue" />
    );

    expect(container.firstChild).toMatchSnapshot();
    expect(getByText('TestTitle')).toHaveStyleRule('margin-bottom', '5px');
    expect(getByText('TestValue')).toHaveStyleRule('font-size', '1.3rem');
  });

  test('renders correctly when `bold` prop passed in', () => {
    const { container, getByText } = render(
      <AccountDetail title="TestTitle" value="TestValue" bold={true} />
    );

    expect(container).toMatchSnapshot();
    expect(getByText('TestValue')).toHaveStyleRule('font-size', '1.5rem');
    expect(getByText('TestValue')).toHaveStyleRule('font-weight', 'bold');
  });

  test('renders correctly when `inline` prop passed in', () => {
    const { container, getByText } = render(
      <AccountDetail title="TestTitle" value="TestValue" inline={true} />
    );

    expect(container).toMatchSnapshot();
    expect(getByText('TestTitle')).toHaveStyleRule('margin-right', '15px');
    expect(getByText('TestValue')).toHaveStyleRule('display', 'inline-block');
    expect(getByText('TestValue')).toHaveStyleRule('display', 'inline-block');
  });
});
