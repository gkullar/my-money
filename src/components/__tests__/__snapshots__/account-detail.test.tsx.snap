// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountDetail renders correctly 1`] = `
.c0 {
  font-size: 0.75rem;
  text-transform: uppercase;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  margin-bottom: 5px;
}

.c1 {
  font-size: 1.3rem;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
}

<div
  class="account-detail"
>
  <div
    class="c0"
  >
    TestTitle
  </div>
  <div
    class="c1"
  >
    TestValue
  </div>
</div>
`;

exports[`AccountDetail renders correctly when \`bold\` prop passed in 1`] = `
.c0 {
  font-size: 0.75rem;
  text-transform: uppercase;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  margin-bottom: 5px;
}

.c1 {
  font-size: 1.3rem;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  font-weight: bold;
  font-size: 1.5rem;
}

.c2 {
  background-color: #ff5248;
  color: #1b1b1b;
  box-shadow: 8px 8px 4px -4px rgba(255,82,72,0.4);
  border-radius: 2px;
  padding: 15px;
  text-align: center;
  position: fixed;
  width: 300px;
  margin-left: -150px;
  left: 50%;
  bottom: 30px;
  z-index: 1;
  visibility: hidden;
}

@media (min-width:30em) {
  .c2 {
    width: 400px;
    margin-left: -200px;
  }
}

<div>
  <div
    class="account-detail"
  >
    <div
      class="c0"
    >
      TestTitle
    </div>
    <div
      class="c1"
    >
      TestValue
    </div>
  </div>
  <div
    class="c2"
  />
</div>
`;

exports[`AccountDetail renders correctly when \`inline\` prop passed in 1`] = `
.c0 {
  font-size: 0.75rem;
  text-transform: uppercase;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  display: inline-block;
  margin-right: 15px;
}

.c1 {
  font-size: 1.3rem;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  display: inline-block;
}

.c2 {
  background-color: #ff5248;
  color: #1b1b1b;
  box-shadow: 8px 8px 4px -4px rgba(255,82,72,0.4);
  border-radius: 2px;
  padding: 15px;
  text-align: center;
  position: fixed;
  width: 300px;
  margin-left: -150px;
  left: 50%;
  bottom: 30px;
  z-index: 1;
  visibility: hidden;
}

@media (min-width:30em) {
  .c2 {
    width: 400px;
    margin-left: -200px;
  }
}

<div>
  <div
    class="account-detail"
  >
    <div
      class="c0"
    >
      TestTitle
    </div>
    <div
      class="c1"
    >
      TestValue
    </div>
  </div>
  <div
    class="c2"
  />
</div>
`;
